### PR TITLE
ka - reimplements effectiveCapacity issue

### LINKED ISSUE #8: FEATURE: As an admin, I can specify a capacity "per farmer"
# User Story

* As an admin
* I can specify a commons capacity on a per farmer basis
* so that the game will scale with the number of users automatically

Part 1 in #28 
Part 2 in #29 
Part 3 in #33 

### Discussion

Prof. de Vries would like to be able to specify commons capacity on a "per user" basis.

The idea is to add a field `capacityPerUser` that will be multiplied by the number of users (farmers) in the commons.    

Then, when doing capacity calculations for the health update formulas, the capacity used will be `max(carryingCapacity, numUsers * capacityPerUser)`.

### Acceptance Criteria

- [ ] There is a way for the admin to set capacityPerUser in addition to carryingCapacity; and both fields appear when you create or edit a commons.
- [ ] On the `/admin/listcommons` page, you can see both the carrying capacity, the perUserCapacity, and the effectiveCapacity, which is `max(carryingCapacity, numUsers * capacityPerUser)`
- [ ] The effectiveCapacity is used in update health formulas instead of the carryingCapacity.

### Implementation Todos

# Fix backend crud operations

- [ ] The Commons entity has a new field for capacityPerUser
- [ ] The POST operation for creating a new commons can set this field
- [ ] The PUT operation for editing a commons can update this field
- [ ] The commons entity has a getter for effectiveCapacity which return `max(carryingCapacity, numUsers * capacityPerUser)`
- [ ] The commons entity has annotation to return the value of this serialized in the JSON.

# Fix health operations
- [ ] The health formulas are updated to use effectiveCapacity

# Fix frontend
- [ ] Frontend create and edit pages for commons provide access to capacityPerUser
- [ ] List Commons page for admins shows capacityPerUser and effectiveCapacity in addition to all current fields.

### LINKED ISSUE #29: FEATURE: As an admin, I have access to an effectiveCapacity variable (Break from #8)
# User Story

* As an admin
* I can specify a commons capacity on a per farmer basis
* so that the game will scale with the number of users automatically

Part 2 of issue #8. In this issue, the focus is on implementing a fully functional ```effectiveCapacity``` field.
Continuation of issue #28 and must be completed in order.

### Acceptance Criteria

- [x] On the `/admin/listcommons` page, you can see ALL THREE the carrying capacity, the perUserCapacity, and the effectiveCapacity, which is `max(carryingCapacity, numUsers * capacityPerUser)`

### Implementation Todos

# Fix backend crud operations
- [x] The commons entity has a getter for effectiveCapacity which return `max(carryingCapacity, numUsers * capacityPerUser)`
- [x] The commons entity has annotation to return the value of this serialized in the JSON.

# Fix frontend
- [x] List Commons page for admins shows BOTH capacityPerUser and effectiveCapacity.

### LINKED ISSUE #33: FEATURE: Health formulas use effectiveCapacity (Break from #8)
# User Story

* As an admin
* I can specify a commons capacity on a per farmer basis
* so that the game will scale with the number of users automatically

Part 3 of issue #8. In this issue, the focus is on actually using the ```effectiveCapacity``` field in the formulas.
Continuation of issue #29 and must be completed in order.

### Acceptance Criteria
- [ ] The effectiveCapacity is used in update health formulas instead of the carryingCapacity.

### Implementation Todos

# Fix health operations
- [ ] The health formulas are updated to use effectiveCapacity

### PR DESCRIPTION
**Pull request by kirrarista**
Implements front- and backend for the Effective Capacity field allowing for dynamic adjustment of carrying capacity based on the number of players.

image

Combines the methods Rocky used in #59 and #60 with instructor provided feedback.

Deployed for testing on https://proj-happycows-kirrarista.dokku-05.cs.ucsb.edu/

Closes #29
Closes #33
Closes #8

### CODE REVIEW

**Review by jczhou02 — APPROVED** [review]
Looks good to me!

**Review by pconrad — APPROVED** [review]
LGTM!

File: frontend/src/main/components/Commons/CommonsTable.js [hunk 1]
```diff
@@ -35,8 +35,6 @@ export default function CommonsTable({ commons, currentUser }) {
         navigate(`/leaderboard/${cell.row.values["commons.id"]}`)
     }
 
-    const downloadCallback = (cell) => window.location.href = `/api/commons/${cell.row.values["commons.id"]}/download?commonsId=${cell.row.values["commons.id"]}`;
-
```
Comment (chriscerie) [inline frontend/src/main/components/Commons/CommonsTable.js:39]: I'm confused why is download being reverted?

**Review by chriscerie — CHANGES_REQUESTED** [review]

File: frontend/src/main/components/Commons/CommonsTable.js (hunk 1 above)
Comment (pconrad) [reply frontend/src/main/components/Commons/CommonsTable.js:39]: Good catch.  I'm asking the team now

File: frontend/src/main/components/Commons/CommonsTable.js [hunk 2]
```diff
@@ -95,17 +93,20 @@ export default function CommonsTable({ commons, currentUser }) {
...
 
     const columnsIfAdmin = [
         ...columns,
-        ButtonColumn("Edit","primary", editCallback, testid),
-        ButtonColumn("Delete","danger", deleteCallback, testid),
-        ButtonColumn("Leaderboard","secondary", leaderboardCallback, testid),
-        ButtonColumn("Download", "success", downloadCallback, testid)
+        ButtonColumn("Edit", "primary", editCallback, testid),
+        ButtonColumn("Delete", "danger", deleteCallback, testid),
+        ButtonColumn("Leaderboard", "secondary", leaderboardCallback, testid)
```
Comment (pconrad) [inline frontend/src/main/components/Commons/CommonsTable.js:109]: same question here.. why does this PR remove the downloads?

File: frontend/src/tests/components/Commons/CommonsTable.test.js [hunk 3]
```diff
@@ -145,29 +141,4 @@ describe("UserTable tests", () => {
 
   });
 
-  test("the download button works as intended", async () => {
```
Comment (pconrad) [inline frontend/src/tests/components/Commons/CommonsTable.test.js:148]: Why ???

File: frontend/src/main/components/Commons/CommonsTable.js (hunk 1 above)
Comment (pconrad) [reply frontend/src/main/components/Commons/CommonsTable.js:39]: Should these lines be restored?

File: frontend/src/main/components/Commons/CommonsTable.js [hunk 4]
```diff
@@ -95,17 +93,20 @@ export default function CommonsTable({ commons, currentUser }) {
...
     ];
 
     const testid = "CommonsTable";
 
     const columnsIfAdmin = [
         ...columns,
-        ButtonColumn("Edit","primary", editCallback, testid),
-        ButtonColumn("Delete","danger", deleteCallback, testid),
-        ButtonColumn("Leaderboard","secondary", leaderboardCallback, testid),
-        ButtonColumn("Download", "success", downloadCallback, testid)
```
Comment (pconrad) [inline frontend/src/main/components/Commons/CommonsTable.js:108]: Should this be restored?

File: frontend/src/tests/components/Commons/CommonsTable.test.js [hunk 5]
```diff
@@ -109,7 +106,6 @@ describe("UserTable tests", () => {
     expect(screen.getByTestId(`${testId}-cell-row-0-col-Edit-button`)).toHaveClass("btn-primary");
     expect(screen.getByTestId(`${testId}-cell-row-0-col-Delete-button`)).toHaveClass("btn-danger");
     expect(screen.getByTestId(`${testId}-cell-row-0-col-Leaderboard-button`)).toHaveClass("btn-secondary");
-    expect(screen.getByTestId(`${testId}-cell-row-0-col-Download-button`)).toHaveClass("btn-success");
```
Comment (pconrad) [inline frontend/src/tests/components/Commons/CommonsTable.test.js:112]: Should this be restored?

File: frontend/src/tests/components/Commons/CommonsTable.test.js (hunk 3 above)
Comment (pconrad) [reply frontend/src/tests/components/Commons/CommonsTable.test.js:148]: Should this entire test be restored?

**Review by pconrad — CHANGES_REQUESTED** [review]
Can you either (a) restore the download feature or (b) explain why this PR, in particular, is removing it?

**Review by pconrad — APPROVED** [review]
LGTM

**Pull request merged**